language/tools/move-package/src/compilation/compiled_package.rs

```diff
@@ -143,24 +137,37 @@ impl OnDiskCompiledPackage {
         );
         let compiled_units = compiled_units
             .iter()
-            .zip(source_maps.iter())
-            .map(|(bytecode_path, source_map_path)| {
-                let bytecode_bytes = std::fs::read(bytecode_path.as_str())?;
-                let source_map = source_map_from_file(Path::new(source_map_path))?;
+            .map(|bytecode_path| {
+                let bytecode_path = Path::new(bytecode_path.as_str());
+                let file_stem = bytecode_path.file_stem().unwrap();
+                let bytecode_bytes = std::fs::read(&bytecode_path)?;
+                let source_map = source_map_from_file(
+                    &CompiledPackageLayout::SourceMaps
+                        .from_sibling_path(bytecode_path)
+                        .ok_or_else(|| anyhow::format_err!("Unable to find source map"))?
```

Comment:
So does this mean that source map files are required to be included with Move bytecode? Or am I misunderstanding? I had always thought of source map files as optional, so this surprises me a little, but I admit I don't know a ton about their history of use.

Reply:
Until now source maps have been largely viewed as optional, but I don't think that is what we want in the package system. The main reason for  this is in order for a number of tools to have good experiences (or to be usable at all) we need the source maps (source coverage, disassembler, bytecode viewer, unit testing, prover). Keeping these additional artifacts around as required is fine here, since the package system takes the view that it is the sole creator and manager of Move bytecode/Move-related artifact, and the compiler produces them for us anyways -- we've just been ignoring them until now. 

So for "normal packages" source maps are not considered optional. We may in the future want to support "binary packages" that don't contain source maps or other rich information, but that you can link/compile against though (this would also be useful for compiling against artifacts on the chain).

Reply:
Awesome, that all sounds great! I am definitely supportive of *always* tracking what Move source location corresponds to each bytecode instruction (in all contexts where Move source exists, that is). It sounds like requiring source maps in more and more parts of the Move toolchain is a good way to progress toward that goal.

Given that, I'm curious if you have any thoughts on somehow embedding source locations into the bytecode, through some sort of "bytecode metadata." It seems like right now we have separate files: an `.mvsm` for every `.mv`. Is that a conscious, permanent choice? I could definitely see how that's a big advantage when it comes to keeping bytecode files small (sort of how [split DWARF](https://www.productive-cpp.com/improving-cpp-builds-with-split-dwarf/) allows debug information -- source locations, essentially -- to be split out of object files).

But, on the other hand, if we're moving toward a world that increasingly requires source mappings to be present, storing that information in separate files means we're always going to have to write logic for the case where we can find the `.mv` file on disk, but can't find the corresponding `.mvsm`. Do you envision a future where all of this information is stored in a `.mv` file, and "binary packages" are essentially the same thing but with the metadata stripped out? Or is that something you're explicitly hoping to avoid (perhaps for size/performance reasons)?

Reply:
I'm still figuring out what we want long-term here, but at least at the moment, I tend to be in the "keep them split" camp. In particular, I like them being separate from performance/size perspective -- if compiled modules are what are published and executed on chain, then I think keeping these as minimal as possible (within reason) makes sense.

Additionally, the source maps only really make sense if you also have the source file (at least in most cases), so coupling them to the bytecode doesn't make as much sense -- really you want a coupling of source file, source map, and bytecode to get the best experience, but publishing the source code on chain feels like that isn't optimal either.